zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -93,35 +109,100 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
     }
 
     // ============================================================================================ //
+    // Reminder: The primary inputs are either field elements, hash digest or binary arrays.
+    // As a field element may not represent an entire hash digest, we formatted the latter in two parts:
+    // a field element and aggregated the remaining bits into (a) field element(s): the residual field element(s).
+    // For efficiency we also added in these residual field elements the public values (bit arrays of 64 bits).
+    //
     // Reminder: Remember that the primary inputs are ordered as follows:
     // We make sure to have the primary inputs ordered as follow:
-    // [Root, NullifierS, CommitmentS, value_pub_in, value_pub_out]
+    // [Root, NullifierS, CommitmentS, h_sig, h_iS, Residual Field Element(S)]
     // ie, below is the index mapping of the primary input elements on the protoboard:
     // - Index of the "Root" field elements: {0}
     // - Index of the "NullifierS" field elements: [1, NumInputs + 1[
     // - Index of the "CommitmentS" field elements: [NumInputs + 1, NumOutputs + NumInputs + 1[
-    // - Index of the "v_pub_in" field element: {NumOutputs + NumInputs + 1}
-    // - Index of the "v_pub_out" field element: {NumOutputs + NumInputs + 1 + 1}
+    // - Index of the "h_sig" field element: {NumOutputs + NumInputs + 1}
+    // - Index of the "Message Authentication TagS" (h_i) field elements: [NumOutputs + NumInputs + 1 + 1, NumOutputs + NumInputs + 1 + NumOuputs [
+    // - Index of the "Residual Field Element(s)" field elements: [NumOutputs + NumInputs + 1 + NumOuputs + 1 , NumOutputs + NumInputs + 1 + NumOuputs + 1 + nb_field_residual]
+    //
+    // The Residual field elements are structured as follows:
+    // - v_pub_in [0, size_value[
+    // - v_pub_out [size_value + 1, 2*size_value[
+    // - h_sig remaining bits [2*size_value + 1, 2*size_value + (digest_length-field_capacity)[
+    // - nullifierS remaining bits [2*size_value + (digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs)*(digest_length-field_capacity)[
+    // - commitmentS remaining bits [2*size_value + (1+NumInputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity)[
+    // - message authentication tagS remaining bits [2*size_value + (1+NumInputs+NumOutputs)*(digest_length-field_capacity) + 1, 2*size_value + (1+2*NumInputs+NumOutputs)*(digest_length-field_capacity)]
     // ============================================================================================ //
 
+    // This function extract the remaining bits of the nullifierS, commitmentS or h_iS from the residual field element(S)
+    function extract_extra_bits(uint start, uint end, uint[] memory primary_inputs) public pure returns (bytes1) {
```

Comment:
Please provide tests for the newly introduced functions

Reply:
Sure. 

We do not have any test files for the functions in BaseMixer.sol so far. I'll create a specific one for this.